cmd/signing-server/main.go

```diff
@@ -69,14 +71,89 @@ type Config struct {
 	DisableAuth        bool
 	DisableHTTPS       bool
 
+	HSMModule     string
+	HSMTokenLabel string
+	HSMSlot       int
+	HSMPass       string
+	HSMKeyLabel   string
+	HSMKeyId      string
+
+	HSMContext *sign.HSMContext
+
 	// calculated by program
 	Logger *zap.Logger
 }
 
+func (c *Config) SetupHSM() error {
+	c.Logger.Info("setup HSM signing", zap.String("module", c.HSMModule))
+
+	if c.HSMKeyLabel != "" {
+		c.Logger.Info("selecting key label", zap.String("label", c.HSMKeyLabel))
+	}
+	if c.HSMKeyId != "" {
+		c.Logger.Info("selecting key id", zap.String("id", c.HSMKeyId))
+	}
+
+	var slot *int
+	if c.HSMSlot >= 0 {
+		slot = &c.HSMSlot
+	}
+	config := &crypto11.Config{
+		Path:       c.HSMModule,
+		TokenLabel: c.HSMTokenLabel,
+		Slot:       slot,
+		Pin:        c.HSMPass,
+	}
+
+	ctx, err := crypto11.NewSession(config)
+	if err != nil {
+		return err
+	}
+
+	k, err := ctx.FindPrivateKey(c.HSMKeyId, c.HSMKeyLabel)
+	if err != nil {
+		return fmt.Errorf("lookup private key: %w", err)
+	}
+
+	c.HSMContext = sign.NewHSMContext(ctx, k)
+	return nil
```

Comment:
Is this session ever closed?

Reply:
No, it is closed, when the process is done.